radix-engine/src/kernel/call_frame.rs

```diff
@@ -8,29 +8,64 @@ use radix_engine_interface::api::field_api::LockFlags;
 use radix_engine_interface::types::{LockHandle, NodeId, SubstateKey};
 use radix_engine_store_interface::db_key_mapper::SubstateKeyContent;
 
-use super::actor::{Actor, BlueprintHookActor, FunctionActor, MethodActor};
 use super::heap::{Heap, HeapOpenSubstateError, HeapRemoveModuleError, HeapRemoveNodeError};
 use super::kernel_api::LockInfo;
 
 /// A message used for communication between call frames.
 ///
 /// Note that it's just an intent, not checked/allowed by kernel yet.
 #[derive(Default, Debug, Clone, PartialEq, Eq)]
-pub struct Message {
-    pub copy_references: Vec<NodeId>,
+pub struct CallFrameMessage {
+    /// Nodes to be moved from src to dest
     pub move_nodes: Vec<NodeId>,
+
+    /// Copy of a global ref from src to dest
+    pub copy_global_references: Vec<NodeId>,
+
+    /// Copy of a direct access ref from src to dest
+    pub copy_direct_access_references: Vec<NodeId>,
+
+    /// Create a "stable" transient in dest from src. The src node may
+    /// have global or borrowed visibility
+    /// TODO: Cleanup abstraction (perhaps by adding another type of visibility)
+    pub copy_to_stable_transient_references: Vec<NodeId>,
 }
 
-impl Message {
-    pub fn from_indexed_scrypto_value(value: &IndexedScryptoValue) -> Self {
+impl CallFrameMessage {
+    pub fn from_input<C: CallFrameReferences>(value: &IndexedScryptoValue, references: &C) -> Self {
+        let mut copy_global_references = Vec::new();
+        let mut copy_direct_access_references = Vec::new();
+
+        for arg_ref in value.references().clone() {
+            if arg_ref.is_global() {
+                copy_global_references.push(arg_ref);
+            } else {
+                copy_direct_access_references.push(arg_ref);
+            }
+        }
+
+        copy_global_references.extend(references.global_references());
+        copy_direct_access_references.extend(references.direct_access_references());
+
         Self {
-            copy_references: value.references().clone(),
             move_nodes: value.owned_nodes().clone(),
+            copy_global_references,
+            copy_direct_access_references,
+            copy_to_stable_transient_references: references.stable_transient_references(),
+        }
+    }
+
+    pub fn from_output(value: &IndexedScryptoValue) -> Self {
+        Self {
+            copy_global_references: value.references().clone(),
```

Comment:
Should split `value.references()` in case a function returns a direct access references.

Reply:
good catch